### ISSUE TITLE
Update protocol to include message length as first part

### ISSUE THREAD

**Issue by yasserf** [deepstreamIO/deepstream.io#184]
Following the great feedback by @Matthias247 in #104 it would be worth visiting whether we want to introduce a message length prefix for all the messages we send:

```
E|S|bob+
```

Would be:

```
7|E|S|bob+
```

**Comment by KrishnaPG** [comment]
The idea in #104 is `against using special characters for termination of fields` and rather `length prefixing` them - and not a combination of *both* as seems to be indicated by `7|E|S|bob+` 

----------------------------

-  It is unnecessarily complex to compute the length and "prefix" it in languages like C/C++ (suffixing the length is relatively easier, but it does not serve much purpose). 

As for other languages (such as Javascript, which seems to be the predominant users), the length is not really needed upfront (at the application level). 

Besides, in the C/C++ arena this especially affects the zero-copy and streaming performance code paths.

For example, in C/C++ it is currently very easy to **reuse** the `received` buffer to send back `ack` or `rej` just by changing few characters in-place and reusing the buffer (without having to do a fresh allocation and copying the whole content over). Note that this zero-copy is order of magnitudes faster thanks to consistent request-response message formats. 

Length prefixes greatly alter this (e.g. REQ message lengt may not be same as REJ length).

----------------

- On the other hand, not sure what exact purpose the prefixed-length solves in either language, if you want to **retain** the special characters (as seems to be suggested by the example `7|E|S|bob+` )

What is really needed is `start positions` of various parts of the message, something like:
````
P|REQ|toUppercase|<UID>|Sabc+
````
would be
````
[5,0,1,4,15,20]
PREQtoUppercase<UID>Sabc+
````
where the `[5,0,1,4,15,20]` indicates that the message has `5` parts each stating at `0,1,4,15 and 20th` index respectively. (This is just an example, and in reality you may want to add the length of "[5,0,1,4,15,20]" also to the indices, or send the length and data separately etc...)


The bottom line is `special characters` + `prefix length` does not any problem, and only hinders the performance. Only one of them is needed, not both.

**Comment by yasserf** [comment]
#796
